lib/dangermattic/plugins/android_unit_test_pr_checker.rb

```diff
@@ -0,0 +1,128 @@
+    #
+    # @example Check missing unit tests with a custom bypass label
+    #   check_missing_tests(bypass_label: 'BypassTestCheck')
+    def check_missing_tests(classes_exceptions: CLASSES_EXCEPTIONS, subclasses_exceptions: SUBCLASSES_EXCEPTIONS,
+                            bypass_label: UNIT_TESTS_BYPASS_PR_LABEL)
+      list = find_classes_missing_tests(
+        git_diff: git.diff,
+        classes_exceptions: classes_exceptions,
+        subclasses_exceptions: subclasses_exceptions
+      )
+
+      return if list.empty?
+
+      if danger.github.pr_labels.include?(bypass_label)
+        list.each do |c|
+          warn("Class `#{c.classname}` is missing tests, but `#{bypass_label}` label was set to ignore this.")
+        end
+      else
+        list.each do |c|
+          failure("Please add tests for class `#{c.classname}` (or add `#{bypass_label}` label to ignore this).")
+        end
+      end
+    end
+
+    private
+
+    ClassViolation = Struct.new(:classname, :file)
+
+    # @param [Git::Diff] git_diff the object
+    # @return [Array<ClassViolation>] An array of `ClassViolation` objects for each added class that is missing a test
+    def find_classes_missing_tests(git_diff:, classes_exceptions:, subclasses_exceptions:)
+      violations = []
+      removed_classes = []
+      added_test_lines = []
+
+      # Parse the diff of each file, storing test lines for test files, and added/removed classes for non-test files
+      git_diff.each do |file_diff|
+        path = file_diff.path
+        if test_file?(path: path)
+          # Store added test lines from test files
+          added_test_lines += file_diff.patch.each_line.select do |line|
+            GitUtils.change_type(diff_line: line) == :added
+          end
+        else
+          # Detect added and removed classes in non-test files
+          file_diff.patch.each_line do |line|
+            case GitUtils.change_type(diff_line: line)
+            when :added
+              matches = line.scan(NON_PRIVATE_CLASS_DETECTOR)
+              matches.reject! do |m|
+                class_match_is_exception?(
+                  match: m,
+                  file: path,
+                  classes_exceptions: classes_exceptions,
+                  subclasses_exceptions: subclasses_exceptions
+                )
+              end
+              violations += matches.map { |m| ClassViolation.new(m[0], path) }
+            when :removed
+              matches = line.scan(ANY_CLASS_DETECTOR)
+              removed_classes += matches.map { |m| m[0] }
+            end
+          end
+        end
+      end
+
+      # We only want newly added classes, not if class signature was modified or line was moved
+      violations.reject! { |v| removed_classes.include?(v.classname) }
+      # For each remaining candidate, only keep the ones _not_ used in a new test
+
+      violations.select { |v| added_test_lines.none? { |line| line =~ /\b#{v.classname}(Tests?)?\b/ } }
```

Comment:
Ah! I misunderstood the intention here and then that is the reason for a few other points you mentioned below. I thought it was looking for actual test class counterparts of the original classes (which would make the sole word `Classname` a bit odd, but I thought that was the intention if for some reason we had a test class inside a test folder with the exact same name, but it could be a project convention or something like that), not uses of that class.

> We don't want a violation to be excluded if the PR introduces a new class named Foo inside Foo.kt, and then introduces a new file FeatureTests.kt whose content includes class FooTests {} but without any actual test calling/referencing the Foo class itself (like new Foo())

Clear!

Reply:
Maybe we could add some `# code comment` above that `\b#{v.classname\b` regex to clarify this for future readers of the code and avoid future confusion?